core/trino-main/src/main/java/io/trino/operator/window/RowsFraming.java

```diff
@@ -0,0 +1,144 @@
+        }
+        else if (frameInfo.getStartType() == PRECEDING) {
+            frameStart = preceding(rowPosition, getValue(frameInfo.getStartChannel(), currentPosition));
+        }
+        else if (frameInfo.getStartType() == FOLLOWING) {
+            frameStart = following(rowPosition, endPosition, getValue(frameInfo.getStartChannel(), currentPosition));
+        }
+        else {
+            frameStart = rowPosition;
+        }
+
+        // frame end
+        if (frameInfo.getEndType() == UNBOUNDED_FOLLOWING) {
+            frameEnd = endPosition;
+        }
+        else if (frameInfo.getEndType() == PRECEDING) {
+            frameEnd = preceding(rowPosition, getValue(frameInfo.getEndChannel(), currentPosition));
+        }
+        else if (frameInfo.getEndType() == FOLLOWING) {
+            frameEnd = following(rowPosition, endPosition, getValue(frameInfo.getEndChannel(), currentPosition));
+        }
+        else {
+            frameEnd = rowPosition;
+        }
+
+        return new Range(frameStart, frameEnd);
+    }
+
+    private boolean emptyFrame(FrameInfo frameInfo, int rowPosition, int endPosition)
+    {
+        FrameBound.Type startType = frameInfo.getStartType();
+        FrameBound.Type endType = frameInfo.getEndType();
+
+        int positions = endPosition - rowPosition;
+
+        if ((startType == UNBOUNDED_PRECEDING) && (endType == PRECEDING)) {
+            return getValue(frameInfo.getEndChannel(), 0) > rowPosition;
+        }
+
+        if ((startType == FOLLOWING) && (endType == UNBOUNDED_FOLLOWING)) {
+            return getValue(frameInfo.getStartChannel(), 0) > positions;
+        }
+
+        if (startType != endType) {
+            return false;
+        }
+
+        FrameBound.Type type = frameInfo.getStartType();
+        if ((type != PRECEDING) && (type != FOLLOWING)) {
+            return false;
+        }
+
+        long start = getValue(frameInfo.getStartChannel(), 0);
+        long end = getValue(frameInfo.getEndChannel(), 0);
+
+        if (type == PRECEDING) {
+            return (start < end) || ((start > rowPosition) && (end > rowPosition));
+        }
+
+        return (start > end) || (start > positions);
+    }
+
+    private static int preceding(int rowPosition, long value)
+    {
+        if (value > rowPosition) {
+            return 0;
+        }
+        return toIntExact(rowPosition - value);
+    }
+
+    private static int following(int rowPosition, int endPosition, long value)
+    {
+        if (value > (endPosition - rowPosition)) {
+            return endPosition;
+        }
+        return toIntExact(rowPosition + value);
+    }
+
+    private long getValue(int channel, int currentPosition)
+    {
+        checkState(!pagesIndex.isNull(channel, currentPosition), "Window frame offset must not be null");
+        long value = pagesIndex.getLong(channel, currentPosition);
+        checkState(value >= 0, "Window frame offset must not be negative");
+        return value;
+    }
```

Comment:
Apparently, you didn't hit any test with this changed error message. We might be missing something.

Reply:
Ah, it's a checkstate... The actual check is done elsewhere. But a test is missing anyway.